File: 백준/Gold/3584. 가장 가까운 공통 조상/가장 가까운 공통 조상.js
const fs = require("fs");
const [firstLine, ...input] = fs
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const t = Number(firstLine);
const numbers = input.map((str) => str.split(" ").map(Number));

function solution(n, a, b, vertex) {
  const tree = Array.from({ length: n + 1 }, () => new Array());
  const parents = Array.from({ length: n + 1 }, () => 0);
  const children = Array.from({ length: n + 1 }, () => 0);

  vertex.forEach(([x, y]) => {
    parents[y] = x;
    children[x]++;
    tree[x].push(y);
  });

  let root = vertex[0][0];
  while (parents[root] !== 0) {
    root = parents[root];
  }

  const levels = Array.from({ length: n + 1 }, () => 0);
  const queue = [root];
  let front = 0;
  while (front !== queue.length) {
    const cur = queue[front++];
    tree[cur].forEach((child) => {
      levels[child] = levels[cur] + 1;
      queue.push(child);
    });
  }

  while (a !== b) {
    const depthA = levels[a];
    const depthB = levels[b];
    a = depthA >= depthB ? parents[a] : a;
    b = depthA <= depthB ? parents[b] : b;
  }
  return a;
}

for (let i = 0; i < t; i++) {
  const [[n]] = numbers;
  const [, ...vertex] = numbers.splice(0, n + 1);
  const [[a, b]] = vertex.splice(-1);
  console.log(solution(n, a, b, vertex));
}

File: 백준/Gold/3584. 가장 가까운 공통 조상/가장 가까운 공통 조상.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int solve(int n);

int main() {
  int t, n;
  cin>>t;
  while(t--) {
    cin>>n;
    cout<<solve(n)<<'\n';
  }
  return 0;
}

int solve(int n) {
  vector <int> v[10001];
  int levels[10001] = {0, }, children[10001] = {0, }, parents[10001] = {0, };

  int x, y, a, b;
  for(int i = 1; i < n; i++) {
    cin>>x>>y;
    v[x].push_back(y);
    parents[y] = x;
    children[x]++;
  }
  cin>>a>>b;

  int root = x;
  while(parents[root] != 0) {
    root = parents[root];
  }

  queue <int> q;
  q.push(root);
  int cur, child;
  while(!q.empty()) {
    cur = q.front();
    q.pop();
    for(int i = 0; i < v[cur].size(); i++) {
      child = v[cur][i];
      levels[child] = levels[cur] + 1;
      q.push(child);
    }
  }

  while(a != b) {
    if(levels[a] > levels[b]) {
      a = parents[a];
    } else if(levels[a] < levels[b]) {
      b = parents[b];
    } else {
      a = parents[a];
      b = parents[b];
    }
  }
  return a;
}